File: Problems/Leetcode/3001-4000/3201-3300/3249.cpp
#include <bits/stdc++.h>

using namespace std;
using ll = long long;

static constexpr int MOD = 1e9 + 7;

class Solution {
    vector<int> C;

    int dfs(int x, int pa, vector<vector<int>>& g) {
        int cnt = 1;
        for (int y : g[x]) {
            if (y != pa) {
                cnt += dfs(y, x, g);
            }
        }
        return C[x] = cnt;
    }

    int dfs2(int x, int pa, vector<vector<int>>& g) {
        set<int> s;
        int res = 0;
        for (int y : g[x]) {
            if (y != pa) {
                s.emplace(C[y]);
                res += dfs2(y, x, g);
            }
        }
        if (s.size() <= 1) {
            res++;
        }
        return res;
    }

public:
    int countGoodNodes(vector<vector<int>>& edges) {
        int n = edges.size() + 1;
        vector<vector<int>> g(n);
        for (auto& e : edges) {
            int u = e[0], v = e[1];

            g[u].emplace_back(v);
            g[v].emplace_back(u);
        }

        C.resize(n, 0);
        dfs(0, -1, g);

        return dfs2(0, -1, g);
    }
};

int main() {
    Solution sol;

    return 0;
}

File: Problems/Leetcode/3001-4000/3201-3300/3249.py
from typing import List


class Solution:
    def countGoodNodes(self, edges: List[List[int]]) -> int:
        n = len(edges) + 1
        g = [[] for _ in range(n)]
        for u, v in edges:
            g[u].append(v)
            g[v].append(u)

        ans = 0

        def dfs(u: int, pa: int = -1):
            size, pre = 1, 0
            ok = 1
            for v in g[u]:
                if v == pa:
                    continue
                sz = dfs(v, u)
                if pre != 0 and sz != pre:
                    ok = 0
                size += sz
                pre = sz

            nonlocal ans
            ans += ok
            return size

        dfs(0)
        return ans
